### PR TITLE
Move new/edit external attachment page to the GOV.UK Design System

### PR DESCRIPTION
**Pull request by davidtrussler**
The work in this PR moves the add/edit external attachment pages to the design system layout. See the [Trello card](https://trello.com/c/UUh0Bjp9/612-move-new-edit-attachment-page-to-the-govuk-design-system) for this work. 

| Current layout | Design System layout |
|---|---|
| Screenshot 2022-08-31 at 08 36 52 | Screenshot 2022-09-01 at 15 02 40 |
|  | Screenshot 2022-09-01 at 15 08 37 |

### CODE REVIEW

**Comment by nnagewad** [comment]
Looks good. Just checking is the bottom margin between inputs 50px?

**Comment by nnagewad** [comment]
Also, just double check the spacing for the back button and the sub-title. May be good to reference what @DilwoarH has done on the taxonomy page.

**Comment by davidtrussler** [comment]
@nnagewad I haven't set any specific spacing on these elements. I'll cross reference with the work that's already been done on this as you suggest in the first instance.

File: app/views/admin/attachments/edit.html.erb
Comment (DilwoarH) [inline app/views/admin/attachments/edit.html.erb:5]: If you add it like this, it will add everything you need

Comment (DilwoarH) [reply app/views/admin/attachments/edit.html.erb:5]:
```ruby
<% content_for :back_link do %>
  <%= render "govuk_publishing_components/components/back_link", {
    href: admin_edition_path(@edition)
  } %>
<% end %>
```

File: app/views/admin/attachments/edit.html.erb
Comment (DilwoarH) [inline app/views/admin/attachments/edit.html.erb:1]: You probably want to follow this pattern:

```
<% content_for :page_title, "Editing #{attachment.attachable_model_name} #{attachment.readable_type} attachment" %>
<% content_for :title, "Edit #{attachment.readable_type} attachment" %>
<% content_for :context, "Attachments for #{attachment.attachable_model_name}" %>
```

File: app/views/admin/attachments/edit_legacy.html.erb
Comment (DilwoarH) [inline app/views/admin/attachments/edit_legacy.html.erb:1]: same as https://github.com/alphagov/whitehall/pull/6763/files#r960413584

**Comment by davidtrussler** [comment]
@nnagewad I've updated the spacing between elements on this to be 50px each time. I also updated how the back button and heading are rendered, which is via a standard method across all these pages. I've updated the screenshots to show this. 

The effect of this is to render the relative sizes of the headings (the main one and the ones for each element) now appear to be the wrong size relative to each other. I initially followed the font sizes in the Figma file but could yo confirm what we would expect these to be? Currently they are:

|Element|font-size|Figma font-size|
|---|---|---|
|"Attachments for publication"|24px|27px|
|"Add external attachment"|36px|48px|
|"Title (required)"|36px|36px|

**Comment by nnagewad** [comment]
Hey David, regarding the size of the font title, I was using Extra large font size + caption as a reference.
- Link to [GOV.UK Typography](https://design-system.service.gov.uk/styles/typography/#headings-with-captions)

File: app/views/admin/attachments/edit.html.erb
Comment (davidtrussler) [reply app/views/admin/attachments/edit.html.erb]: Nice, have done.

File: app/views/admin/attachments/edit.html.erb
Comment (davidtrussler) [reply app/views/admin/attachments/edit.html.erb]: 👍

File: app/views/admin/attachments/edit_legacy.html.erb
Comment (davidtrussler) [reply app/views/admin/attachments/edit_legacy.html.erb:1]: 👍

**Comment by davidtrussler** [comment]
This was updated in the [last commit](https://github.com/alphagov/whitehall/pull/6763/commits/97cb5b9dab28682e20cfb3589ebd97cda23b5f50) so that input, radio and select elements persist the form values where applicable if the user submits the form with errors. Some of this looks a bit clunky to me and there’s probably a more elegant way of achieving this. 

I was also looking for a way to replicate the current behaviour of highlighting which inputs are causing the errors but was unable to see how that can be done. 

I am wondering if this port to the design system, which effectively bypasses the form builder currently in use is going to be problematic and where we should look at ways of porting it whilst maintaining the form builder.

File: app/views/admin/attachments/_form.html.erb
Comment (DilwoarH) [inline app/views/admin/attachments/_form.html.erb:6]: Is this the correct variable to use?

Comment (DilwoarH) [reply app/views/admin/attachments/_form.html.erb:6]: What does it output?

File: app/views/admin/attachments/_form.html.erb
Comment (DilwoarH) [inline app/views/admin/attachments/_form.html.erb:22]: I think you can just pass in `form.object.title` and it will handle the empty in the component.

```suggestion
      value: form.object.title
```

File: app/views/admin/attachments/_form.html.erb
Comment (DilwoarH) [inline app/views/admin/attachments/_form.html.erb:59]:
```suggestion
        value: form.object.external_url
```

File: app/views/admin/attachments/_form.html.erb
Comment (DilwoarH) [inline app/views/admin/attachments/_form.html.erb:73]: We should probably replace this with govuk buttons

File: app/views/admin/attachments/_reference_fields.html.erb
Comment (DilwoarH) [inline app/views/admin/attachments/_reference_fields.html.erb:15]:
```suggestion
    value: form.object.isbn
```

File: app/views/admin/attachments/_reference_fields.html.erb
Comment (DilwoarH) [inline app/views/admin/attachments/_reference_fields.html.erb:27]:
```suggestion
    value: form.object.unique_reference
```

File: features/step_definitions/attachment_steps.rb
Comment (DilwoarH) [inline features/step_definitions/attachment_steps.rb:34]: instead of a new step, we can just do this:

```
@user.has_permission? "Preview design system"
```

File: features/support/attachment_helper.rb
Comment (DilwoarH) [inline features/support/attachment_helper.rb:35]:
```suggestion
    design_system = @user.has_permission? "Preview design system"
    click_on design_system ? "Next" : "Save"
```

File: features/support/attachment_helper.rb
Comment (DilwoarH) [inline features/support/attachment_helper.rb:26]:
```suggestion
  def create_external_attachment(url, attachment_title)
```

File: features/support/attachment_helper.rb
Comment (DilwoarH) [inline features/support/attachment_helper.rb:43]:
```suggestion
    create_external_attachment("http://www.example.com/example", "Example doc")
```

File: features/step_definitions/attachment_steps.rb
Comment (DilwoarH) [inline features/step_definitions/attachment_steps.rb:33]:
```suggestion
```

File: features/step_definitions/attachment_steps.rb
Comment (DilwoarH) [inline features/step_definitions/attachment_steps.rb:29]:
```suggestion
  create_external_attachment(url, title)
```

File: features/edition-attachments.feature
Comment (DilwoarH) [inline features/edition-attachments.feature:33]:
```suggestion
    And I add an external attachment with the title "Beard Length Illustrations" and the URL "http://www.beardlengths.gov.uk"
```

File: app/views/admin/attachments/_form.html.erb
Comment (davidtrussler) [reply app/views/admin/attachments/_form.html.erb:6]: Good question, I have no idea where that came from. It outputs the type of attachment (e.g. "external"). I've replaced it with what we've used elsewhere i.e. `#{attachment.readable_type}`

File: app/views/admin/attachments/_form.html.erb
Comment (davidtrussler) [reply app/views/admin/attachments/_form.html.erb]: 👍

File: app/views/admin/attachments/_form.html.erb
Comment (davidtrussler) [reply app/views/admin/attachments/_form.html.erb]: 👍

File: app/views/admin/attachments/_form.html.erb
Comment (davidtrussler) [reply app/views/admin/attachments/_form.html.erb:71]: Yes, I believe that's correct: it's the equivalent of the "save" button that we have currently.

File: app/views/admin/attachments/_form.html.erb
Comment (davidtrussler) [reply app/views/admin/attachments/_form.html.erb:73]: I've just kept the existing buttons here for now as it's out of scope for external attachments. This will be replaced when the other forms of attachments are being worked on.

File: app/views/admin/attachments/_reference_fields.html.erb
Comment (davidtrussler) [reply app/views/admin/attachments/_reference_fields.html.erb]: 👍

File: app/views/admin/attachments/_reference_fields.html.erb
Comment (davidtrussler) [reply app/views/admin/attachments/_reference_fields.html.erb]: 👍

File: features/step_definitions/attachment_steps.rb
Comment (davidtrussler) [reply features/step_definitions/attachment_steps.rb]: Ah right, nice!

File: features/support/attachment_helper.rb
Comment (davidtrussler) [reply features/support/attachment_helper.rb]: 👍

File: features/support/attachment_helper.rb
Comment (davidtrussler) [reply features/support/attachment_helper.rb]: 👍

File: features/step_definitions/attachment_steps.rb
Comment (davidtrussler) [reply features/step_definitions/attachment_steps.rb]: 👍

File: features/edition-attachments.feature
Comment (davidtrussler) [reply features/edition-attachments.feature]: 👍

**Comment by davidgisbey** [comment]
@davidtrussler  I'll whip up an error component that links to the inputs etc. and push it up on a different branch. If you're happy you can just cherry pick the commit 👍

**Comment by davidtrussler** [comment]
> @davidtrussler I'll whip up an error component that links to the inputs etc. and push it up on a different branch. If you're happy you can just cherry pick the commit 👍

@davidgisbey Sounds good to me! 👍

File: app/views/admin/attachments/_reference_fields_legacy.html.erb
Comment (davidgisbey) [inline app/views/admin/attachments/_reference_fields_legacy.html.erb:4]: This section needs to be updated to use the design systems select component

image

File: app/views/admin/attachments/_form.html.erb
Comment (DilwoarH) [inline app/views/admin/attachments/_form.html.erb:10]: Minor styling changes:

```suggestion
      items: form.object.errors.full_messages.map do |msg|
        {
          text: msg
        }
      end
```

File: app/views/admin/attachments/_form.html.erb
Comment (DilwoarH) [inline app/views/admin/attachments/_form.html.erb:72]: Did some new logic get added here? Looking at the legacy file, it seems like the save_or_cancel always shows.

File: app/views/admin/attachments/_hoc_reference_fields.html.erb
Comment (DilwoarH) [inline app/views/admin/attachments/_hoc_reference_fields.html.erb:25]: I think you can just put this in:

```suggestion
        checked: form.object.numbered_hoc_paper,
```

File: app/views/admin/attachments/_hoc_reference_fields.html.erb
Comment (DilwoarH) [inline app/views/admin/attachments/_hoc_reference_fields.html.erb:36]:
```suggestion
        checked: form.object.unnumbered_hoc_paper,
```

File: app/views/admin/attachments/_reference_fields.html.erb
Comment (DilwoarH) [inline app/views/admin/attachments/_reference_fields.html.erb:8]:
```suggestion
  options.concat(attachable.translated_locales.map do |locale|
    {
      text: native_language_name_for(locale),
      value: locale
    }
  end)
```

File: features/support/attachment_helper.rb
Comment (DilwoarH) [inline features/support/attachment_helper.rb:32]: I think the convention for test files is as follows:

def create_external_attachment(url, attachment_title)
  design_system = @user.has_permission? "Preview design system"

  click_on "Add new external attachment"
  fill_in "Title", with: attachment_title
  fill_in "External url", with: URL
  click_on design_system ? "Next" : "Save"
end

**Comment by DilwoarH** [comment]
Hey @davidtrussler - looks good generally, some minor spacing issues but seems to be okay.

Would be useful to jump on a call to see a demo of this.

File: app/views/admin/attachments/_form.html.erb
Comment (davidtrussler) [reply app/views/admin/attachments/_form.html.erb:10]: 👍

File: app/views/admin/attachments/_form.html.erb
Comment (davidtrussler) [reply app/views/admin/attachments/_form.html.erb:72]: The idea is to keep `save_or_cancel` when the attachment is anything other than an external attachment, in which case it uses the design system `next` button. Either way it submits the form.

File: app/views/admin/attachments/_hoc_reference_fields.html.erb
Comment (davidtrussler) [reply app/views/admin/attachments/_hoc_reference_fields.html.erb:25]: Hmm, I tried some variations on that but it didn't persist the state of the radio when the user submits the form with errors. The problem is that `numbered_hoc_paper` is not a valid thing though `unnumbered_hoc_paper` is. I am aware though that setting the values the way it is done here looks pretty clunky and I'm sure there's a more elegant way to deal with it. I just couldn't see what it would be.

**Comment by davidtrussler** [comment]
> Hey @davidtrussler - looks good generally, some minor spacing issues but seems to be okay.
> 
> Would be useful to jump on a call to see a demo of this.

Yes, a demo would be good. Still need to properly sort out the checked value for the radios.

File: features/support/attachment_helper.rb
Comment (davidtrussler) [reply features/support/attachment_helper.rb]: OK, makes sense 👍

File: app/views/admin/attachments/_hoc_reference_fields.html.erb
Comment (davidtrussler) [reply app/views/admin/attachments/_hoc_reference_fields.html.erb]: I have now updated this to something that seems to render as required and is somewhat more elegant.

File: app/views/admin/attachments/_reference_fields.html.erb
Comment (davidtrussler) [reply app/views/admin/attachments/_reference_fields.html.erb]: 👍

File: app/views/admin/attachments/_form.html.erb
Comment (DilwoarH) [reply app/views/admin/attachments/_form.html.erb:74]: I think we should add in the styling here because everything else is using govuk design system except this:

```erb
<div class="govuk-button-group">
  <%= render "govuk_publishing_components/components/button", {
    text: "Save"
  } %>

  <a class="govuk-link" href="<%= attachable_attachments_path(attachable) %>">Cancel</a>
</div>
```

File: app/views/admin/attachments/_hoc_reference_fields.html.erb
Comment (DilwoarH) [inline app/views/admin/attachments/_hoc_reference_fields.html.erb:35]: Minor nit pick

```suggestion
      selected: (form.object.parliamentary_session == session)
```

File: app/views/admin/attachments/_reference_fields.html.erb
Comment (DilwoarH) [inline app/views/admin/attachments/_reference_fields.html.erb:60]:
```suggestion
        checked: form.object.unnumbered_command_paper.eql?(false),
```

**Review by DilwoarH — CHANGES_REQUESTED** [review]
Looks great - some minor changes but once these are done, happy to approve

File: app/views/admin/attachments/_hoc_reference_fields.html.erb
Comment (davidtrussler) [reply app/views/admin/attachments/_hoc_reference_fields.html.erb:12]: Oh yeah, of course it does. 👍

File: app/views/admin/attachments/_hoc_reference_fields.html.erb
Comment (davidtrussler) [reply app/views/admin/attachments/_hoc_reference_fields.html.erb:35]: Yeah that's better 👍

File: app/views/admin/attachments/_form.html.erb
Comment (DilwoarH) [inline app/views/admin/attachments/_form.html.erb:77]: Missing the wrapper

```suggestion
 <div class="govuk-button-group">
    <%= render "govuk_publishing_components/components/button", {
      text: "Save"
    } %>
    <a class="govuk-link" href="attachable_attachments_path(attachable)">Cancel</a>
 </div>
```

**Comment by davidtrussler** [comment]
@DilwoarH The error messages have been added to the form fields as below in the last commit. There is also one further commit to persist the values for the numbers for the papers when the form is submitted with errors. 

Screenshot 2022-09-14 at 15 32 24

File: app/controllers/admin/attachments_controller.rb
Comment (davidgisbey) [inline app/controllers/admin/attachments_controller.rb:88]: I think this needs to be scoped to the specific controller actions you're working on or it'll apply to all the controller actions (index etc.)

```

  def get_layout
    return "admin" unless preview_design_system_user?

    case action_name
    when "edit", "update", "new", "create"
      "design_system"
    else
      "admin"
    end
  end
```

File: app/views/admin/attachments/_form.html.erb
Comment (davidgisbey) [inline app/views/admin/attachments/_form.html.erb:4]: sorry i should've said that i merged an error summary component which renders at the top of the page and links to the inputs! My fault. You can call it like this 

```
<% content_for :error_summary, render('shared/error_summary', object: attachment.becomes(Attachment), class_name: @attachment.class.name.demodulize.underscore.humanize.downcase) %>

```

File: app/views/admin/attachments/_form.html.erb
Comment (davidgisbey) [inline app/views/admin/attachments/_form.html.erb:20]: So for example here this would need to be 

```
<%= render "govuk_publishing_components/components/input", {
      label: {
        text: "Title (required)"
      },
      name: "attachment[title]",
      id: "attachment[title],
   heading_level: 2,
      heading_size: "l",
      value: form.object.title,
      error_items: (form.object.errors.as_json(full_messages: true)[:title].map do |error|
        {
          "text": error,
        }
      end if form.object.errors[:title].present?)
    } %>
```

for the errors to link

File: app/views/admin/attachments/_form.html.erb
Comment (davidgisbey) [inline app/views/admin/attachments/_form.html.erb:24]: I've added a helper method here for this https://github.com/alphagov/whitehall/blob/ed6c5d2a4a4438387f4d8e4cbfc21ffe937e33cf/app/helpers/errors_helper.rb

You can call it by doing the following

```
errors_for(object.errors, :attribute)
```

So in this case it would be 

```
errors_for(attachment.errors, :title)
```

Comment (davidgisbey) [reply app/views/admin/attachments/_form.html.erb:24]: Yours does something v similar though 👍

File: app/views/admin/attachments/_form.html.erb
Comment (davidgisbey) [inline app/views/admin/attachments/_form.html.erb:66]: same as above regarding helper methods for errors and giving it an id so the error links correctly

**Comment by davidgisbey** [comment]
Looks great. I've left a few comments around using helpers and the error summary component. Let me know if any of it doesn't make sense 👍

File: app/controllers/admin/attachments_controller.rb
Comment (davidtrussler) [reply app/controllers/admin/attachments_controller.rb]: Ah right, makes sense 👍

**Comment by davidtrussler** [comment]
> Looks great. I've left a few comments around using helpers and the error summary component. Let me know if any of it doesn't make sense or you want to pair on it 👍

Cheers @davidgisbey - that's all clear. The only instance that's a bit problematic is on the `<select>` element, which has a slightly different form to `<input>` elements. I've added it [here](https://github.com/alphagov/whitehall/pull/6763/files#diff-05c5cb654af952bfb6350ee3376d3356b359a7d394b86c50fef228864c63af6cR52) which renders as below. I'm not sure if the error data needs to be passed differently or if the helper needs a tweak. 

Screenshot 2022-09-15 at 17 44 02

**Comment by davidgisbey** [comment]
You can use the error_for_input helper method for this (should probs be renamed to errors_for_select as that's the only one that can't take an array of hashes.

So here it would be `errors_for_input(attachment.errors, : parliamentary_session)`

File: app/controllers/admin/attachments_controller.rb
Comment (davidgisbey) [inline app/controllers/admin/attachments_controller.rb:32]: If the fail is unsuccessful it will rerender the edit view here https://github.com/alphagov/whitehall/blob/87af8214f730585c3273613ee4a31f5ad26a06f6/app/controllers/admin/attachments_controller.rb#L44-L46

We will also need this logic 
```
render(preview_design_system_user? ? "edit" : "edit_legacy")
```

there so the correct page is rerendered

File: app/controllers/admin/attachments_controller.rb
Comment (davidgisbey) [inline app/controllers/admin/attachments_controller.rb:19]: If the fail is unsuccessful it will rerender the edit view here https://github.com/alphagov/whitehall/blob/87af8214f730585c3273613ee4a31f5ad26a06f6/app/controllers/admin/attachments_controller.rb#L26-L28

We will also need this logic 
```
 render(preview_design_system_user? ? "new" : "new_legacy")```

there so the correct page is rerendered

File: app/views/admin/attachments/_form.html.erb
Comment (davidtrussler) [reply app/views/admin/attachments/_form.html.erb]: Updated to use the helper.

**Review by DilwoarH — APPROVED** [review]
Looks good - let's wait till Tuesday before we merge as it's a long weekend and Friday afternoon

**Review by davidgisbey — APPROVED** [review]
lgtm

**Pull request merged**
